### ISSUE TITLE
nodeInit() unnecessarily calls DescribeNetworkInterfaces() IMDS more than once

### ISSUE THREAD

**Issue by jaypipes** [aws/amazon-vpc-cni-k8s#808]
nodeInit() - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/ipamd/ipamd.go#L271) function grabs all the ENIs for the worker node, then grabs all pods on the worker node and loops through all the pods setting up iptables rules.

This piece of code: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/ipamd/ipamd.go#L373-L378

can and should be pulled out of that loop since it's not dependent on the contents of the loop iteration. This would eliminate N-1 calls to IMDS to get the subnets associated with the node, where N == #pods

**Comment by jaypipes** [comment]
Actually, it's even less efficient than that... We already call GetVPCIPv4CIDRs() on line 305 and throw away the result :(

https://github.com/aws/amazon-vpc-cni-k8s/blob/master/ipamd/ipamd.go#L305

**Comment by jaypipes** [comment]
nodeInit() is making all these calls to DescribeNetworkInterfaces(), but it's not actually the GetVPCIPv4CIDRs() call that is doing it. It's all the callers of IPAMContext.getENIAddresses(): https://github.com/aws/amazon-vpc-cni-k8s/blob/4bf10a04402b1249f755fff36a79f608136d79ff/ipamd/ipamd.go#L792

IPAMContext.getENIAddresses() calls the awsClient.DescribeENI() method, which bypasses the "cache" and calls the IMDS's DescribeNetworkInterfaces() API call.
[January 28, 2020, 8:48 AM] Pipes, Jay: getENIAddresses() is called from two places: IPAMContext.tryAssignIPs() and IPAMContext.setupENI(). Both of those locations are called in nodeInit(), with the setupENI() called inside a loop across each ENI for the instance type and tryAssignIPs() doing yet another loop across ENIs for the instance type.

In short, the solution here is to call DescribeNetworkInterfaces() *once* at the start of nodeInit() and pass the results of that call along to subroutines like setupENI() and tryAssignIPs().
